proposals/2025-01-02_direct-io.md

```diff
@@ -0,0 +1,166 @@
+    // May trigger an implicit flush if necessary.
+    Write([]byte) (int, error)
+
+    // Flushes any buffered data to the file.
+    // The writer may not be usable after a call to Flush().
+    Flush() error
+
+    // Discards any unflushed buffered data and clears any errors.
+    // Resets the writer to use the specified file.
+    Reset(f *os.File) error
+}
+```
+
+Additional utilities will also need to be developed to:
+
+* Automatically discover alignment requirements (when supported):
+
+  ```go
+  // directIORqmts holds the alignment requirements for direct I/O.
+  // All fields are in bytes.
+  type directIORqmts struct {
+    // The required alignment for memory buffers addresses.
+    memoryAlign int
+    // The required alignment for I/O segment lengths and file offsets.
+    offsetAlign int
+  }
+
+  func fetchDirectIORqmts(fd uintptr) (*directIORqmts, error)
+  ```
+
+* Allocate aligned memory uffers:
+
+  ```go
+  func alignedBlock(size int, alignmentRqmts *directIORqmts) []byte
+  ```
+
+* Enable Direct I/O on a fd:
+
+  ```go
+  func enableDirectIO(fd uintptr) error
+  ```
+
+Direct I/O support and alignment requirementsdiffer across OSes and filesystems. To keep the scope
+manageable, the initial focus will be on adding support for Linux (ideally versions `>6.1`). Build
+tags will enfore this restriction. Attempting to enable the feature flag on an unsupported OS will
+result in am error.
+
+Support for additional operating systems and filesystems will be added subsequently.
+
+Another key consideration is ensuring that all tests, especially for TSDB, are also run with direct
+I/O enabled. A build tag will be added to force the direct I/O writer for testing purposes, allowing
+for commands such as:
+
+```shell
+go test --tags=forcedirectio ./tsdb/
+```
+
+For the direct I/O writer itself and its utils, different unit tests will be added.
+
+However, micro-benchmarking the writer may be misleading , as its performance is highly dependent
+on usage patterns. While some micro-benchmarks will be provided to ensure consistency with
+`bufio.Writer` (without sacrificing performance), higher-level benchmarks which simulate real
+Prometheus processes (such as compaction instead of writing a `10GB` buffer which is unlikely to
+be encountered in Prometheus) will receive greater emphasis. These benchmarks may also help
+fine-tune parameters like the buffer size.
+
+For a look at the outcomes, the screenshot below shows a reduction of `20-50%` in page cache usage,
+as measured by the `container_memory_cache` metric while running Prombench on
+[Prometheus PR #15365](https://github.com/prometheus/prometheus/pull/15365). In this test, the
+direct I/O writer was used for writing chunks during compaction. No significant regressions or
+adverse effects on CPU or disk I/O were observed.
+
+![container_memory_cache](../assets/2025-01-02_direct-io/container_memory_cache.png)
+
+## Alternatives
```

Comment:
Using `posix_fadvise` with `POSIX_FADV_DONTNEED` could be an option if it doesn't make sense to totally avoid the cache (there's also Linux specific memory management options like `MADV_COLD`, which could help for some of the container concerns to optimise which memory is reclaimed, although it wouldn't likely have as much user visible impact on the page cache).

I suspect those doesn't make sense for the TSDB but could be worth mentioning here, there might be other places in the ecosystem where those make more sense than `O_DIRECT`.

Reply:
Thnaks, I'll mention that.

Reply:
Also, work on `RWF_UNCACHED` has resumed https://lore.kernel.org/linux-fsdevel/20241111234842.2024180-1-axboe@kernel.dk/T/#cluster-upstream-ci

Thinking about going with an `use-uncached-io` feature flag, give direct I/O a chance, then reconsider once `RWF_UNCACHED` is usable.